In [2]:
# Imports and pip installations (if needed)
%pip install pandas
%pip install -U numpy scipy scikit-learn
import pandas as pd
from sklearn.preprocessing import Normalizer

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [3]:
# Load the given datasets
df_numerical = pd.read_csv('chronic_disease_numerical.csv') #for numerical dataset
df_categorical = pd.read_csv('chronic_disease_categorical.csv') #for categorical dataset

# Pri(nt the data
print(df_numerical.head(10))
print(df_categorical.head(10))

   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   
5     992643  35.0  70.0   82.0   36.0  1.1  150.0  3.5  14.5  52.0  9400.0   
6     573400  60.0  80.0    NaN    NaN  NaN    NaN  NaN   NaN   NaN     NaN   
7     109053  64.0  60.0  106.0   27.0  0.7  150.0  3.3  14.4  42.0  8100.0   
8     352995  33.0  60.0  130.0   41.0  0.9  141.0  4.4  15.5  52.0  4300.0   
9     395450  65.0  70.0  203.0   46.0  1.4    NaN  NaN  11.4  36.0  5000.0   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  
5   6.1  
6   NaN  
7   4.7  
8   5.8  
9   4.1  
   unique_id   al   

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
The dataset on chronic kidney disease (CKD) from the UCI repository contains 25 attributes or features, including the target class, which categorizes patients as either having CKD or not. These features represent various medical and biological measurements, and they are categorized into numerical and nominal (categorical) types. Here’s a breakdown of the data:

### **Features (Attributes) Explanation:**
- **Age:** Age of the patient in years (Numerical).
- **Blood Pressure:** Blood pressure in mm/Hg (Numerical).
- **Specific Gravity:** A measure of kidney function, indicating how concentrated the urine is. It is recorded as one of the values (1.005, 1.010, 1.015, 1.020, 1.025) (Categorical/nominal).
- **Albumin:** Protein level in the urine, with values from 0 to 5 (Categorical/nominal).
- **Sugar:** Sugar level in the urine, also measured on a scale from 0 to 5 (Categorical/nominal).
- **Red Blood Cells:** This feature indicates whether red blood cells are normal or abnormal in the patient’s urine (Categorical/nominal).
- **Pus Cell:** Indicates whether pus cells in the urine are normal or abnormal (Categorical/nominal).
- **Pus Cell Clumps:** Presence or absence of pus cell clumps in urine (Categorical/nominal).
- **Bacteria:** Presence or absence of bacteria in the urine (Categorical/nominal).
- **Blood Glucose Random:** Random blood glucose level, measured in mg/dl (Numerical).
- **Blood Urea:** The amount of urea in the blood, measured in mg/dl (Numerical).
- **Serum Creatinine:** Creatinine level in the blood, measured in mg/dl (Numerical).
- **Sodium:** Sodium concentration in blood, measured in mEq/L (Numerical).
- **Potassium:** Potassium concentration in blood, measured in mEq/L (Numerical).
- **Hemoglobin:** Hemoglobin level in the blood, measured in grams (Numerical).
- **Packed Cell Volume:** Volume percentage of red blood cells in the blood (Numerical).
- **White Blood Cell Count:** Count of white blood cells per cubic millimeter (cells/cumm) (Numerical).
- **Red Blood Cell Count:** Count of red blood cells per cubic millimeter (millions/cmm) (Numerical).
- **Hypertension:** Whether the patient has hypertension (yes/no) (Categorical/nominal).
- **Diabetes Mellitus:** Whether the patient has diabetes mellitus (yes/no) (Categorical/nominal).
- **Coronary Artery Disease:** Whether the patient has coronary artery disease (yes/no) (Categorical/nominal).
- **Appetite:** Indicates if the patient’s appetite is good or poor (Categorical/nominal).
- **Pedal Edema:** Presence of swelling in the legs (yes/no) (Categorical/nominal).
- **Anemia:** Whether the patient has anemia (yes/no) (Categorical/nominal).
- **Class:** The target label indicating whether the patient has chronic kidney disease (ckd/notckd) (Categorical/nominal).

### **Features vs. Labels:**
- **Features:** All the attributes except the last one (Class) are features. These include a mix of biological and medical measurements.
- **Label (Target):** The "Class" column represents whether the patient is diagnosed with CKD (chronic kidney disease) or not (ckd/notckd).

### **Feature Types:**
- **Numerical Features:** These features are continuous or discrete numbers that represent measurements in various units.
- **Categorical (Nominal) Features:** These features are categorical and represent classifications or binary options.

### **Examples of Numerical Columns:**
1. **Age** – Patient’s age in years.
2. **Blood Urea** – Blood urea concentration in mg/dl.
3. **Sodium** – Sodium level in mEq/L.

### **Examples of Categorical (Nominal) Columns:**
1. **Specific Gravity** – Different levels of urine concentration (1.005, 1.010, etc.).
2. **Red Blood Cells** – Whether red blood cells in the urine are normal or abnormal.
3. **Hypertension** – Whether the patient has hypertension (yes/no).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates_numerical = df_numerical.duplicated().sum()
print("Number of duplicated rows in numerical dataset: ", duplicates_numerical)
# Drop these duplicate rows
df_numerical.drop_duplicates(inplace=True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates_categorical = df_categorical.duplicated().sum()
print("Number of duplicated rows in categorical dataset: ", duplicates_categorical)
# Drop these duplicate rows
df_categorical.drop_duplicates(inplace=True)

Number of duplicated rows in numerical dataset:  30
Number of duplicated rows in categorical dataset:  30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

In [5]:
# Checking the shape of the two datasets, observing the difference in the shape of both datasets to confirm the 
# mismatch in row counts
print("Rows in categorical dataset", df_categorical.shape[0])
print("Rows in numerical dataset", df_numerical.shape[0])

# Perform a set difference check to help identify rows that exist in one dataset but are missing from the other
missing_in_categorical = df_numerical[~df_numerical['unique_id'].isin(df_categorical['unique_id'])]
print("Missing in categorical dataset\n", missing_in_categorical)
missing_in_numerical = df_categorical[~df_categorical['unique_id'].isin(df_numerical['unique_id'])]
print("Missing in numerical dataset\n", missing_in_numerical)

Rows in categorical dataset 400
Rows in numerical dataset 390
Missing in categorical dataset
 Empty DataFrame
Columns: [unique_id, age, bp, bgr, bu, sc, sod, pot, hemo, pcv, wbcc, rbcc]
Index: []
Missing in numerical dataset
      unique_id   al   su       rbc        pc         pcc          ba  htn  \
36      208034  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
58      266335  0.0  0.0    normal    normal  notpresent  notpresent   no   
75      325238  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
125     240430  0.0  0.0    normal    normal         NaN         NaN   no   
153     292312  0.0  0.0    normal    normal  notpresent  notpresent   no   
259     298297  0.0  0.0    normal    normal  notpresent  notpresent   no   
279     948317  0.0  0.0    normal    normal  notpresent  notpresent   no   
302     488947  3.0  0.0  abnormal  abnormal  notpresent  notpresent   no   
327     663387  0.0  0.0       NaN    normal  notpresent  notpresent   no   
382 

Answer: 
### **Key Observations:**
**Numerical Dataset**: Has 400 rows after removing duplicates.
**Categorical Dataset**: Has 390 rows after removing duplicates.
**Difference in Row Counts**: There's a discrepancy of 10 rows, meaning there are 10 more rows in the numerical dataset than in the categorical dataset.

### **This difference suggests that not all unique IDs (or rows) in the numerical dataset have corresponding entries in the categorical dataset, or vice versa. There could be several reasons for this:**
**Incomplete Data**: Some rows might be missing from one dataset due to incomplete data collection.
**Data Entry Errors**: It’s possible that some IDs or rows were mistakenly left out or recorded incorrectly, leading to mismatches.
**Filtering or Outliers**: Certain rows could have been filtered out during data cleaning or preprocessing, especially if they contained outliers or were deemed unreliable.

### **A combined table would look like:**
When combining the two datasets (categorical + numerical) based on the unique IDs, the rows that are missing from the numerical dataset would show up in the combined table with empty or null values for the numerical columns, because I'll be using an outer join in this case. In other words, those rows would have values for all the categorical attributes, but their corresponding numerical columns would contain NaN (Not a Number) or missing data markers.
The case wherein I didn't specify the usage of any joins in df.merge, I only got 390 rows that is only the rows with matching IDs from the other dataset were included.


In [6]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df_combined_datasets = pd.merge(df_numerical, df_categorical, how = 'outer', on="unique_id")

#Print the combined dataset
print("The combined dataset is\n", df_combined_datasets)

The combined dataset is
      unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [8]:
# Calculate the percentage of rows that contain atleast one missing value
missing_val_rows = df_combined_datasets[df_combined_datasets.isna().any(axis=1)]
percentage = (len(missing_val_rows)/len(df_combined_datasets))*100

# Print %
print(percentage)

# Drop these rows from the dataset
df_combined_datasets.dropna(inplace=True)

# Print the Dataset
print(df_combined_datasets)

0.0
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm

## Part 3.4: Sort the dataset according to the Labels

In [9]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df_combined_datasets.sort_values(by=["Target"], axis=0, ascending=True, inplace=True)
df_combined_datasets.reset_index(drop=True, inplace=True)

# Print the dataset
print(df_combined_datasets)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [10]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_combined_datasets = pd.get_dummies(df_combined_datasets, categorical_columns, drop_first=False)

# Print the dataset
print(df_combined_datasets)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

In the downstream modeling process, whether you use dummy variables (multiple columns) or a single binary column depends on the type of categorical variable:

#### **Binary Variables (Two Values):**
If the categorical variable is binary (e.g., 'rbc' with values 'normal' and 'abnormal'), using two dummy columns(`rbc_normal`, `rbc_abnormal`) or a single column (e.g., `rbc` as 0 or 1) would be mathematically equivalent for many models, such as linear regression, logistic regression, and tree-based models. 

#### **Effect of Single Column for Binary Variables:**
In practice, it's common to reduce binary variables to a single column (0 or 1), as this saves memory and computation. For example:
- `rbc` could be represented as 1 for "normal" and 0 for "abnormal".
- This doesn't lose any information because having two columns would be redundant: if `rbc_normal` is 1, you know `rbc_abnormal` must be 0.

For most machine learning models, this would have no effect on predictive performance, and the interpretation remains straightforward. It’s more efficient because you’re not adding unnecessary columns.

### **Categorical Variables with More Than Two Values:**
For categorical variables with more than two unique values (e.g., "type of condiment" with values like `ketchup`, `mustard`, `mayo`, `bbq`, and `ranch`), the situation changes.

#### **Effect of Single Column for Multiclass Variables:**
If you try to use a single column for multiclass categorical variables, this column would typically contain integer values (e.g., 0 for ketchup, 1 for mustard, 2 for mayo, etc.). However, this approach would not be suitable for most models because:
1. Ordinal Interpretation: A model might treat the values as having an ordinal relationship (e.g., `mayo > mustard > ketchup`), even though the categories are unrelated. This could lead to incorrect model assumptions.
2. Loss of Information: Assigning a single integer value to each category prevents the model from understanding that the different categories are independent of each other (i.e., they are not ordered).

Therefore, for categorical variables with more than two unique values, you typically use one-hot encoding (i.e., creating a separate dummy column for each unique value). This way, the model treats each category as independent, without assuming any implicit order.

#### **Example with "Type of Condiment":**
Let’s say the categorical variable "condiment" has five unique values:
- `condiment`: ketchup, mustard, mayo, bbq, ranch.

Using one-hot encoding, you would get five separate columns:
- `condiment_ketchup`
- `condiment_mustard`
- `condiment_mayo`
- `condiment_bbq`
- `condiment_ranch`

Each row would have one column set to 1 (the actual condiment used), and the rest would be 0.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [11]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for cols in numerical_columns:
  mean = df_combined_datasets[cols].mean()
  std_dev = df_combined_datasets[cols].std()
  upper_limit = mean + 3 * std_dev
  lower_limit = mean - 3 * std_dev
  df_combined_datasets = df_combined_datasets[(df_combined_datasets[cols] >= lower_limit) & 
  (df_combined_datasets[cols] <= upper_limit)]

# Print the dataset
print(df_combined_datasets)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [12]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_combined_datasets[numerical_columns] = scaler.fit_transform(df_combined_datasets[numerical_columns])

print(df_combined_datasets[numerical_columns])

          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc      rbcc  
2    0.000000  0.032258  0.395161  0.057143  
5    0.524752  0.548387  0.4435

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: `unique_id` column in this dataset is not appropriate for modeling and prediction. My reasons for this include:
1. No Predictive Power: The unique ID is a purely identifying feature and does not have any correlation with the target variable (i.e., it doesn't help in predicting outcomes like whether a person has chronic kidney disease or not).
2. Irrelevant Information: Including the unique ID in the model can confuse some algorithms, especially in models like linear regression, where numerical values are treated as having a potential relationship with the output. Since the ID is arbitrary, it has no meaningful connection to the prediction.
3. Risk of Overfitting: If the unique ID is left in the dataset, some models (particularly those sensitive to numerical values) might try to find patterns in the ID column that don’t actually exist, leading to overfitting.

In [13]:
#Remove that column
df_combined_datasets.drop(columns=['unique_id'], inplace = True, axis=1)

# Print the dataset
print(df_combined_datasets)

          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
2    0.000000  0.032258  0.395161  ...   False     T

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [14]:
# Export the dataframe to a new csv file
df_combined_datasets.to_csv('chronicDiseaseCombinedDataset.csv')

# Export the dataframe to a new json file
df_combined_datasets.to_json('chronicDiseaseCombinedDataset.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

```markdown
| age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------------|------|---------------|---------------|---------------|---------------|---------------|----------------|----------------|----------------|----------------|---------------|-------|--------------|------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.0 | 0.653226 | 0.257143 | 4.0 | 0.0 | FALSE | TRUE | TRUE | FALSE | FALSE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.905405 | 1.0 | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.0 | 0.148515 | 0.225806 | 0.217742 | 0.057143 | 3.0 | 2.0 | TRUE | FALSE | TRUE | FALSE | FALSE | TRUE | TRUE | FALSE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.743243 | 0.5 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.0 | 0.032258 | 0.395161 | 0.057143 | 2.0 | 0.0 | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.729730 | 0.75 | 0.150215 | 0.281046 | 0.234568 | 0.533333 | 0.793103 | 0.336634 | 0.322581 | 0.5 | 0.314286 | 2.0 | 0.0 | TRUE | FALSE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.540541 | 0.0 | 0.399142 | 0.535948 | 0.358025 | 0.7 | 0.379310 | 0.207921 | 0.161290 | 0.830645 | 0.057143 | 1.0 | 0.0 | FALSE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548 | 0.169355 | 0.114286 | 2.0 | 0.0 | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.716216 | 0.5 | 1.0 | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.387097 | 0.532258 | 0.371429 | 1.0 | 0.0 | TRUE | FALSE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | FALSE | TRUE | TRUE | FALSE | TRUE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.729730 | 0.0 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.064516 | 0.879032 | 0.0 | 3.0 | 1.0 | FALSE | TRUE | TRUE | FALSE | FALSE | TRUE | TRUE | FALSE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.0 | 0.0 | 0.103004 | 0.372549 | 0.074074 | 0.500000 | 0.689655 | 0.217822 | 0.225806 | 1.0 | 0.514286 | 4.0 | 0.0 | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.878378 | 0.0 | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.689655 | 0.366337 | 0.387097 | 0.879032 | 0.371429 | 4.0 | 0.0 | FALSE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE | TRUE | FALSE |
| 0.851351 | 0.25 | 0.618026 | 0.562092

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Language models can sometimes introduce errors or "hallucinations" when converting or interpreting data, as seen above where the LLM couldn't form a markdown table for all of the 15 rows. These hallucinations occur because the model may attempt to infer missing or ambiguous details, resulting in fabricated or incorrect information. Here’s how to mitigate such issues:
1. Cross-Validation with Original Data: Always compare the model's output (converted data) against the original data. Ensure that the numbers, text, or structures match exactly.
2. Manual Spot-Checking: After using an LLM for conversions, manually review a subset of the data. For example, in the top 15 rows, you can manually check whether 3-5 rows are accurately represented.
3. Use Small Batches for Verification: If you’re dealing with a large dataset, break it down into smaller batches (e.g., 10-20 rows) and ask the model to convert one batch at a time. This way, it’s easier to spot and fix potential errors.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [16]:
# Code to rename all the columns in the dataset

# Dictionary mapping abbreviations to full names
column_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'Target': 'CKD/Not-CKD'  
}

# Rename columns in the DataFrame
df_combined_datasets = df_combined_datasets.rename(columns=column_mapping)

# Print the updated dataset to confirm
print(df_combined_datasets.head())


        Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
2  0.743243            0.50              0.442060    0.901961   
5  0.770270            1.00              0.901288    0.163399   
7  0.905405            0.50              0.785408    0.862745   
8  0.202703            0.75              0.158798    0.196078   
9  0.783784            1.00              0.399142    0.287582   

   Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  \
2          0.432099  0.500000   0.793103    0.000000            0.032258   
5          0.345679  0.766667   0.206897    0.524752            0.548387   
7          0.518519  0.600000   1.000000    0.277228            0.322581   
8          0.160494  0.166667   0.206897    0.059406            0.000000   
9          0.839506  0.666667   0.586207    0.019802            0.096774   

   White Blood Cell Count  ...  cad_no  cad_yes  appet_good  appet_poor  \
2                0.395161  ...   False     True       False        True   
5 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [17]:
# Converted SQL to Pandas code
result = df_categorical.groupby('Target').size().reset_index(name='count')
# Display the result
print(result)

   Target  count
0     ckd    248
1   ckd\t      2
2  notckd    150


The Pandas code groups the data in the 'Target' column of df_categorical by each unique value (e.g., 'ckd', 'notckd'). It then counts how many times each unique value appears in the dataset. The final output is a table showing each distinct value in the 'Target' column alongside the number of occurrences (the count).